### PR TITLE
17 add species list predictor into model class

### LINKED ISSUE #17: Add species list predictor into the model class
**Is your feature request related to a problem? Please describe.**
The current model classes for the birdnetlib models do not use the species predictor system it comes with. This should be integrated to enable full usage. 

**Describe the solution you'd like**
Add the species prediction based on longitude and latitude and add logic that calls it if no list is provided but longitude and latitude are.

**Describe alternatives you've considered**
Other than leaving it out and forcing users to provide a list of allowed species, non. 
Seems not a good alernative though.

### PR DESCRIPTION
**Pull request by MaHaWo**
closes #17

### CODE REVIEW

**Comment by MaHaWo** [comment]
# Summary 

- conceptualized the species presence predictor as a kind of model, because that's what it is ultimately. 
- went for the most modular approach, i.e., species list predictor has a default base class `SpeciesPredictorBase` that can be derived from to put in custom solutions, new models or whatever, in the same vane as fof the `model`.
- demand that encapsulate the species presence predictor is bundled with the model if it is present. Birdnet only can use it with the default model, not with custom-trained classifiers.  
- added this class `SpeciesPredictorBase` which derives from `birdnetlib.species.SpeciesList`, overrides the model loading, adds a cache system to automatically cache and retrieve created species lists (lives in OS_cache_dir/iSparrow, but could also live in iSparrow_home/data?), and which simplifies the arguments (only date accepted as arguemnt, not week). 
- added respective unit tests for `SpeciesPredictorBase` and for its integration into the recording class. 
- extend existing tests a bit for individual cases 
- add a security check that checks whether the model, preprocessor and possibly speciespredictor are internally consistend based on the nametags they are given. Needs to be documented somewhere, probably in the sphinx docs? 
- test folder became a bit messy, so reorganized configs and test results into their own folders . 
- did away with relative imports where possible. 
- remove `species_list_path` as argument in the models and the respective tests. this is handled by the cache system of the speciespredictor now.

**Comment by MaHaWo** [comment]
ok, this is the last big thing in this milestone afais @iulusoy.

File: src/iSparrow/sparrow_recording.py [hunk 1]
```diff
@@ -72,6 +119,66 @@ def filename(self):
...
+        """
+        set_analyzer Replace the internally used model and preprocessor.
+
+        Args:
+            model (ModelBase): The classifier model to use
+            preprocessor (PreprocessorBase): The preprocessor to use
+        """
+
+        self._check_model_system_viability(
+            (
```
Comment (iulusoy) [inline src/iSparrow/sparrow_recording.py:137]: as sonracloud flagged this, can you make it
```
list(model.name,  preprocessor.name, model.name if species_predictor is None else species_predictor.name, )
```

File: src/iSparrow/sparrow_recording.py [hunk 2]
```diff
@@ -19,49 +21,94 @@ class SparrowRecording(RecordingBase):
...
+        self.analyzer = model
         self.path = path
         p = Path(self.path)
         self.filestem = p.stem
+        self.species_predictor = None
+        self.allowed_species = []
+
+        # make sure that all the system components are compatible. Based on name tags. Still a bit susceptible. Fix?
+
+        self._check_model_system_viability(
```
Comment (iulusoy) [inline src/iSparrow/sparrow_recording.py:71]: can you pass the list explicitly as
```list( model.name, preprocessor.name, model.name if species_predictor is None else species_predictor.name,)
```

File: src/iSparrow/species_predictor.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        if date is not None:
+
+            week = return_week_48_from_datetime(
+                date
+            )  # not optimal, this is called in ´return_results' again.
+
+        # when we want to use cached results, we first check if we computed a list for this location and date before
+        cache_folder = Path(str(latitude) + "_" + str(longitude) + "_" + str(week))
+
+        if self.use_cache:
```
Comment (iulusoy) [inline src/iSparrow/species_predictor.py:162]:
```
if self.use_cache and if (SpeciesPredictorBase.cache_dir / cache_folder / "species_list.txt").exists():
```

File: tests/speciespredictor_test.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    detections = predictor.predict(
+        p_cfg["latitude"], p_cfg["longitude"], date=date, threshold=p_cfg["threshold"]
+    )
+
+    assert detections == detections_should
+
+    assert predictor.read_from_file == False
+
+    date = datetime(day=10, month=6, year=2021)
```
Comment (iulusoy) [inline tests/speciespredictor_test.py:131]: sonarcloud flags `date` as unused?

**Review by iulusoy — APPROVED** [review]
looks good!

**Pull request merged**